modules/ipp/scanner.go

```diff
@@ -126,86 +172,394 @@ func (scanner *Scanner) GetPort() uint {
+	return mediatype == ContentType, nil
 }
 
-// FIXME: Ensure that an actual content type is being used with some library
-func ippInContentType(resp http.Response) bool {
-	return strings.Contains(resp.Header.Get("Content-Type"), ContentType)
+// FIXME: Cleaner to write this code, possibly slower than copy-pasted version
+// FIXME: Quite possibly not easier to read ("What does storeBody do? Where does it store it?")
+// FIXME: Add some error handling somewhere in here, unless errors should just be ignored and we get what we get
+func storeBody(res *http.Response, scanner *Scanner) {
+	b := bufferFromBody(res, scanner)
+	res.BodyText = b.String()
+	if len(res.BodyText) > 0 {
+		m := sha256.New()
+		m.Write(b.Bytes())
+		res.BodySHA256 = m.Sum(nil)
+	}
 }
 
-// FIXME: Change to instead return (*ScanResults, *zgrab2.ScanError)
-// TODO: Doesn't support TLS at all right now
-func (scanner *Scanner) grab(target zgrab2.ScanTarget) (int8, int8, *zgrab2.ScanError) {
-	//FIXME: This is not where this hostname assignment logic should live
-	host := target.Domain
-	if host == "" {
-		host = target.IP.String()
+// TODO: Make sure this it isn't too slow to use this instead of copy-pasting everywhere necessary
+func bufferFromBody(res *http.Response, scanner *Scanner) *bytes.Buffer {
+	b := new(bytes.Buffer)
+	maxReadLen := int64(scanner.config.MaxSize) * 1024
+	readLen := maxReadLen
+	if res.ContentLength >= 0 && res.ContentLength < maxReadLen {
+		readLen = res.ContentLength
+	}
+	io.CopyN(b, res.Body, readLen)
+	res.Body.Close()
+	res.Body = ioutil.NopCloser(b)
+	return b
+}
+
+// FIXME: add proper error handling to this
+// TODO: Support reading from multiple instances of the same attribute in a response
+func readAttributeFromBody(attrString string, body *[]byte) ([][]byte, error) {
+	attr := []byte(attrString)
+	interims := bytes.Split(*body, attr)
+	if len(interims) > 1 {
+		valueTag := interims[0][len(interims[0])-3]
+		var vals [][]byte
+		buf := bytes.NewBuffer(interims[1])
+		for tag, nameLength := valueTag, int16(0); tag == valueTag && nameLength == 0; {
+			var length int16
+			if err := binary.Read(buf, binary.BigEndian, &length); err != nil {
+				//Couldn't read length of content for some reason
+				//return nil, err
+				continue
+			}
+			val := make([]byte, length)
+			if err := binary.Read(buf, binary.BigEndian, &val); err != nil {
+				//Content not read without issue.
+				//vals = append(vals, val)
+				//return vals, err
+				continue
+			}
+			//return &val, nil
+			vals = append(vals, val)
+			if err := binary.Read(buf, binary.BigEndian, &tag); err != nil {
+				//Couldn't read next valueTag
+				//break
+				continue
+			}
+			if err := binary.Read(buf, binary.BigEndian, &nameLength); err != nil {
+				//Couldn't read next nameLength
+				//break
+				// TODO: Conclude correct error behavior for all of these blocks
+				continue
+			}
+		}
+		return vals, nil // FIXME: This should be whatever error occurred during a break case
```

Comment:
In that case, could you do `return vals, <the error>` instead of `break`, since there is no further handling?